### DISCUSSION TITLE
Adding Buttons dynamically

### DISCUSSION THREAD

**Discussion by peter1960** [slint-ui/slint#5362] [Q&A] [answered]
I need to add various buttons/text on start-up based and what was read from an ini file.
How should this be done, I have been looking at various examples but nothing I can see is what am after.

**Comment by ogoffart** [comment]
You can create the buttons with a `for` loop from a model. For example, this slint code:

```slint
import { AboutSlint, Button, VerticalBox } from "std-widgets.slint";
export component Demo {
    in property <[string]> buttons: ["Ok", "Cancel", "Foo", "Bar"];
    callback button-clicked(int);
    button-clicked(idx) => { debug("Button \{idx} was clicked") }

    preferred-height: 300px;

    VerticalBox {
        alignment: end;
        HorizontalLayout {
            alignment: center;
            for btn[index] in buttons: Button { 
                text: btn;
                clicked => { button-clicked(index); }
            }
        }
    }
}
```

And you can et the `buttons`  property from your native code dynamically. And listen to the button-clicked callback

**Reply by Vadoola** [reply, marked as answer]
What @ogoffart suggested is exactly what I'm doing in my program and it works well. My application is themeable and the theme selection buttons are determined at runtime based on the theme files found on disk.

If you want to look at my code for a more detailed example feel free: https://github.com/Vadoola/Tomotroid

Relevant Sections:

- fn load_themes() in [settings.rs](https://github.com/Vadoola/Tomotroid/blob/main/src/settings.rs)
- The following code from [main.rs](https://github.com/Vadoola/Tomotroid/blob/main/src/main.rs)
   `let settings = settings::load_settings();`
    `let theme_model: Rc<VecModel<JsonTheme>> = Rc::new(VecModel::from(themes));`
    `window.global::<ThemeCallbacks>().set_themes(ModelRc::from(theme_model.clone()));`
 -  The [themeconfig.slint](https://github.com/Vadoola/Tomotroid/blob/main/ui/themeconfig.slint) file
 - The section of [main.rs](https://github.com/Vadoola/Tomotroid/blob/main/src/main.rs) with the `.on_theme_changed()` callback (search for it, it's around line 620-625 in the current code, no reason to post the whole code here)

Hope it helps.

**Reply by peter1960** [reply]
This is simple and close to what I need, thanks for sharing and I will do some study
